src/compiler/transformers/automatic-key-insertion/index.ts

```diff
@@ -0,0 +1,136 @@
+import ts from 'typescript';
+
+import { deriveJSXKey } from './utils';
+
+/**
+ * A transformer factory to create a transformer which will add `key`
+ * properties to all of the JSX nodes contained inside of a component's `render`
+ * function.
+ *
+ * This can be thought of as transforming the following:
+ *
+ * ```tsx
+ * class MyComponent {
+ *   render() {
+ *     <div>hey!</div>
+ *   }
+ * }
+ * ```
+ *
+ * to this:
+ *
+ * ```tsx
+ * class MyComponent {
+ *   render() {
+ *     <div key="a-unique-key">hey!</div>
+ *   }
+ * }
+ * ```
+ *
+ * The inserted keys are generated by {@link deriveJSXKey}.
+ *
+ * @param transformCtx a transformation context
+ * @returns a typescript transformer for inserting keys into JSX nodes
+ */
+export const performAutomaticKeyInsertion = (transformCtx: ts.TransformationContext): ts.Transformer<ts.SourceFile> => {
+  /**
+   * This is our outer-most visitor function which serves to locate a class
+   * declaration, at which point it hands things over to the next visitor
+   * function ({@link findRenderMethodVisitor}) which locates the `render`
+   * method.
+   *
+   * @param node a typescript syntax tree node
+   * @returns the result of handling the node
+   */
+  function findClassDeclVisitor(node: ts.Node): ts.VisitResult<ts.Node> {
+    if (ts.isClassDeclaration(node)) {
+      return ts.visitEachChild(node, findRenderMethodVisitor, transformCtx);
+    } else {
+      return ts.visitEachChild(node, findClassDeclVisitor, transformCtx);
```

Comment:
possibly not, just sort of a completeness / correctness thing I guess

but yeah since a class decl will always be a top-level syntax node with a `SourceFile` we can probably just return here, which will cut out some tree-traversal we'd do otherwise

Reply:
I actually found out that we do support non-top-level Stencil components in at least one important case - our own tests files, like here for instance:

https://github.com/ionic-team/stencil/blob/cea23c4a306d05a41f9ca137e9168039276ba30a/src/runtime/test/render-vdom.spec.tsx#L8-L28

I think if we want to continue to support those we'll need to keep traversing the whole tree, unless we want to bite off some more added complexity